File: back-end/src/main/java/project/back_end/service/UserVoucherService.java
package project.back_end.service;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import project.back_end.entity.UserVoucher;
import project.back_end.response.UserVoucherResponse;

@Service
public interface UserVoucherService {

    // Lấy tất cả voucher chưa sử dụng trong ví (trả về Response cho FE)
    Page<UserVoucherResponse> getAvailableVouchers(Long userId, Pageable pageable);

    boolean isVoucherOwned(Long userId, Long voucherId);

    // ================= HÀM NỘI BỘ (Helper) =================
    // Trả về Entity để sử dụng logic nội bộ ở các Service khác (như OrderService)
    UserVoucher getUserVoucherEntity(Long userId, String voucherCode);

    void updateUsageStatus(UserVoucher userVoucher);

    // ================= CRUD CHO ADMIN =================

    // Admin cấp phát mã (trả về Response)
    UserVoucherResponse assignVoucherToUser(Long userId, Long voucherId);

    // Cập nhật: Thêm Pageable, đổi List thành Page
    Page<UserVoucherResponse> getAllUserVouchers(Pageable pageable);

    // Cập nhật: Thêm Pageable, đổi List thành Page
    Page<UserVoucherResponse> getAllVouchersByUserId(Long userId, Pageable pageable);

    void removeVoucherFromUser(Long userVoucherId);
}

File: front-end/src/service/user-voucher-service.ts
import { AxiosResponse } from "axios";
import http from "@/service/http";
import {
    CollectVoucherRequest,
    UserVoucherResponse,
} from "@/type/voucher-type";

// ================= API CHO NGƯỜI DÙNG =================

export const getMyVouchers: () => Promise<
    AxiosResponse<UserVoucherResponse[]>
> = async (): Promise<AxiosResponse<UserVoucherResponse[]>> => {
    return http.get<UserVoucherResponse[]>("/user-vouchers/me");
};

export const collectVoucher: (
    data: CollectVoucherRequest
) => Promise<AxiosResponse<void>> = async (
    data: CollectVoucherRequest
): Promise<AxiosResponse<void>> => {
    return http.post<void>("/user-vouchers/collect", data);
};

// ================= API CHO ADMIN =================

export const getAllUserVouchers: () => Promise<
    AxiosResponse<UserVoucherResponse[]>
> = async (): Promise<AxiosResponse<UserVoucherResponse[]>> => {
    return http.get<UserVoucherResponse[]>("/user-vouchers");
};

export const assignVoucherToUser: (
    userId: number,
    voucherId: number
) => Promise<AxiosResponse<UserVoucherResponse>> = async (
    userId: number,
    voucherId: number
): Promise<AxiosResponse<UserVoucherResponse>> => {
    return http.post<UserVoucherResponse>("/user-vouchers/assign", null, {
        params: { userId, voucherId },
    });
};

export const removeVoucherFromUser: (
    id: number
) => Promise<AxiosResponse<void>> = async (
    id: number
): Promise<AxiosResponse<void>> => {
    return http.delete<void>(`/user-vouchers/${id}`);
};
